### ISSUE TITLE
\TypeError thrown when TwigFunction has an invalid callback.

### ISSUE THREAD

**Issue by menno-ll** [twigphp/Twig#4393]
We're using WordPress and in there we attempt to load Twig templates.
For us we create a WP_CLI command, which registers itself before WordPress itself is loaded. 

Previously (twig version v3.11.1 and before) this all worked great.
Twig allowed functions that were not defined yet to still be compiled as php files.
We did this using the same method as described in https://github.com/twigphp/Twig/issues/3706.
See our [repo of the CLI command](https://github.com/level-level/clarkson-wp-cli-twig-translations/blob/6509c1c52b986be06ca6d82a477e3d42d2ad8dfa/src/GetText.php#L23).
```php
$twig->registerUndefinedFunctionCallback(function ($name) {
    return new \Twig\TwigFunction($name, $name);
});
```

But unfortunately now, since version v3.12.0, this does not work anymore and throws the following TypeError.

```
Uncaught TypeError: Failed to create closure from callable: function "_x" not found or invalid function name in /path/to/project/vendor/twig/twig/src/Util/ReflectionCallable.php:45
```

Now I have to confess i'm not having a full knowledge of the twig codebase yet.
But after searching and looking at the stacktraces, I've found that somewhere around these PR's to be where the problem started: https://github.com/twigphp/Twig/pull/4206 or https://github.com/twigphp/Twig/pull/4207

It seems like even if a function does not exist yet on compilation, it still attempts to extract it's attributes with the new CallableArgumentsExtractor.

Would there be a way to return the functionality to not validate those callable arguments if the function at that stage does not exist, therefore allowing the use of functions that are not yet defined on runtime?

Thanks in advance!

**Comment by stof** [comment]
How would the rendering of your template work if you pass a non-existing PHP callable as the implementation of the Twig function ?

https://github.com/twigphp/Twig/pull/4207 has indeed introduced *more* cases where we inspect the callable with reflection at compile-time. It is now done for every usage of a function, while it was done previously only for calls that involved named arguments.
But this means that we never had a case where we fully supported using an undefined function implementation at compile-time. And it was already broken at runtime previously (as rendering the template would try to call that undefined PHP function).

If your use case is to allow using any function in a template as a no-op (which is pretty weird DX to me), use `return new TwigFunction($name, function () {})` instead.

**Comment by ericmorand** [comment]
@stof so you can't compile a template and then exécute it with a runtime that actually provides the function? What is the benefit of this approach?

**Comment by stof** [comment]
Named arguments in Twig relies on knowing the function signature at compile-time (allowing seamless conversion between snake_case and camelCase argument names would be a no-go if we had to do it at runtime as it would have a performance cost).

**Comment by menno-ll** [comment]
> How would the rendering of your template work if you pass a non-existing PHP callable as the implementation of the Twig function ?

This I can explain.
The command runs without wordpress or plugins loaded.
This would improve performance of the command by quite a lot.
Because simply put, wordpress itself isn't always that performant, and certainly plugins make this way worse.
And for just transforming twig templates to php files, wordpress simply isn't needed.

And because functions that didn't exist at the time of running that command, the twig file would simply be parsed to php code (that indeed at that time contains unexisting functions), this was not a problem.

The generated PHP files were then used by the `wp i18n make-pot` command, which takes php files inside a wordpress theme and attempts to parse php code from them. At that time the functions do exist, as wordpress actually is loaded.

This whole thing is only to just allow wordpress to generate translation templates for translation functions used in twig templates.

> If your use case is to allow using any function in a template as a no-op (which is pretty weird DX to me), use return new TwigFunction($name, function () {}) instead.

I do agree that maybe having wordpress loaded in when running the command would be the best solution, but due to performance reasons i think the fix suggested is way better than what we had.

I guess this solution is way simpler than i imagined!

I'll try it in the following days in some projects and will update this issue if it's indeed fully resolved.

**Comment by ju1ius** [comment]
> we never had a case where we fully supported using an undefined function implementation at compile-time.

@stof See https://github.com/twigphp/Twig/issues/3706#issuecomment-1131854638

> If your use case is to allow using any function in a template as a no-op (which is pretty weird DX to me), use return new TwigFunction($name, function () {}) instead.

The use case is to be able to _defer_ compilation the template to a time where the function is available.
For example you would:
1. On deploy, run your cache warmup cli command without booting your whole application. Which would compile _some_ templates, skipping those with undefined functions.
2. At runtime, when your application is booted, twig can now compile, cache and render the templates used by the current request (or throw an exception if a function is _really_ missing).

Quoting my old self from #3706:
> I know it's not a goot practice, but it is a _very important_ feature when working on legacy applications. It allows integrating twig progressively without requiring a _huge_ amount of work upfront.

**Comment by ju1ius** [comment]
@menno-ll Could you share a minimal reproducer ?

**Comment by stof** [comment]
@ju1ius I would say that you could make your undefined handler callback handler throw a SyntaxError if called during cache warmup (and maybe always return `false` in other cases to let the runtime compilation apply the normal behavior) and make your cache warmup skip templates that trigger a syntax error.

Twig itself cannot implement this "delayed compilation during cache warmup", as it is not responsible for the cache warmup. If you ask Twig to compile a template, it compiles the template. That's all it knows.

**Comment by ju1ius** [comment]
> you could make your undefined handler callback handler throw a SyntaxError if called during cache warmup (and maybe always return false in other cases to let the runtime compilation apply the normal behavior) and make your cache warmup skip templates that trigger a syntax error.

For that, the unknown callback handler would have to be made "aware" of the compilation context...
Unless I'm missing something, that would have to be either:
* supported by the engine, e.g. by passing a compilation context object to every extension
* hacked around via e.g. global or environment variables, which might be okay(ish) in specific contexts but would preclude any kind of reusable solution.

> Twig itself cannot implement this "delayed compilation during cache warmup".

I know. That's not what we ask. We just ask for reflection operations on callables to throw anything that _does not_ extend `Error`, so that the existing behaviour (on which `symfony/twig-bundle` also depends) is preserved.

BTW, I just upgraded Twig to `3.19.0` in an application that depends on this behaviour and things are working fine so far, which is why I asked @menno-ll for a reproducer.

**Comment by ju1ius** [comment]
Looking at this more closely, this is really weird...

The error message in the OP mentions an `uncaught TypeError [...] in ReflectionCallable.php on line 45`, but those are explicitely caught since #3707 ...

**Comment by stof** [comment]
@ju1ius this PR from 2022 does not reflect the state of the code in Twig 3.12+

**Comment by ju1ius** [comment]
> [@ju1ius](https://github.com/ju1ius) this PR from 2022 does not reflect the state of the code in Twig 3.12+

@stof  Yes [it does](https://github.com/twigphp/Twig/blob/v3.19.0/src/Util/ReflectionCallable.php#L45). 😉

**Comment by menno-ll** [comment]
Hi @ju1ius 

Here i have created a repository for you having the issue as described at first.
Inside the readme.md file is described how to reproduce the error.
https://github.com/menno-ll/twig-example

As described in the readme.md as well, the fix from @stof I can confirm after some time of testing works for our scenario.
It seems that this has been implemented in a bad way by us in the first place. It's a miracle this hasn't been a problem for us before. It sill is kind of hacky, but that's okey i guess.

(we still have some other issues to resolve for our projects, but that has nothing to do with this specific issue).

**Comment by stof** [comment]
I'm closing this as you have a working implementation for your hacky case. I don't expect Twig to do changes there to stop using Reflection for those cases. Twig expects a valid callable
